### PR TITLE
cli: make create-plugin prefer version ranges that are already in the lockfile

### LINKED ISSUE #6636: Create-plugin does not respect current app dependencies
## Expected Behavior

The create-plugin command should respect the versions of backstage packages currently installed in the app.

## Current Behavior

Using the create-plugin command installs a new plugin with backstage dependencies that use caret notation, it then runs a yarn install. This bumps backstage packages in the yarn workspace for the entire project and creates mismatches. 

## Possible Solution

Since the backstage versions that get installed are based on the version of the CLI used it's possible to use yarn resolutions to enforce a CLI version (otherwise the CLI version installed in each new plugin will bump the workspace version). However, removing caret notation would potential solve this and allow users to rely on the version bump command to explicitly change versions across the workspace.

## Steps to Reproduce

1. Start with packages/app with not the latest backstage packages
2. Run create-plugin (this may bump some packages, but will certainly bump the CLI)
3. Run create-plugin and note that it's running with the new CLI version and therefore bumping additional packages, potentially even to major versions

## Context

We have external contributors creating plugins as their first and only interaction in the repo. Creating a scenario where they need to be able to adjust other plugins/the app in order to keep backstage package versions consistent is less than ideal. Additionally, this lack of consistency creates errors when running `yarn tsc` in certain cases.

## Your Environment

- Node 14
- yarn workspaces

### PR DESCRIPTION
**Pull request by Rugvip**
Fixes #6636

This should hopefully result in some less noise and fewer touchups needed after using `create-plugin`

### CODE REVIEW

File: packages/cli/src/lib/version.ts [hunk 1]
```diff
@@ -60,3 +62,22 @@ export function findVersion() {
...
+    }
+
+    const lockfileEntries = lockfile?.get(name);
+    if (lockfileEntries?.some(entry => entry.range === '*')) {
+      return '*';
+    }
+    const validRanges = lockfileEntries?.filter(entry =>
+      semver.satisfies(targetVersion, entry.range),
+    );
+    const highestRange = validRanges?.slice(-1)[0];
```
Comment (freben) [inline packages/cli/src/lib/version.ts:80]: are they really guaranteed to be in order?

File: packages/cli/src/lib/version.ts [hunk 2]
```diff
@@ -60,3 +62,22 @@ export function findVersion() {
...
+export function createPackageVersionProvider(lockfile?: Lockfile) {
+  return (name: string, versionHint?: string) => {
+    const targetVersion = versionHint || packageVersions[name];
+    if (!targetVersion) {
+      throw new Error(`No version available for package ${name}`);
+    }
+
+    const lockfileEntries = lockfile?.get(name);
+    if (lockfileEntries?.some(entry => entry.range === '*')) {
+      return '*';
```
Comment (freben) [inline packages/cli/src/lib/version.ts:75]: hm! i would have expected this one to not be "greedy", but rather as a last resort if no actual semver ranges were to be found 🤔

**Review by freben — APPROVED** [review]
Nice, just noob questions

File: packages/cli/src/lib/version.ts (hunk 2 above)
Comment (Rugvip) [reply packages/cli/src/lib/version.ts:75]: The reasoning here is that if there's a `*` it's usually for good reason. Might not always be true though and there might just be some dependency somewhere that has it unnecessarily. I'm thinking we could do this for `@types/*` packages, but not the rest

File: packages/cli/src/lib/version.ts (hunk 1 above)
Comment (Rugvip) [reply packages/cli/src/lib/version.ts:80]: Yeah the lockfile places them in a nice order. The `semver` package doesn't have any range sorting or anything like that so I cba diving into that

**Review by benjdlambert — APPROVED** [review]

**Comment by laurieontech** [comment]
This is great, thanks!

**Pull request merged**